### PR TITLE
DTFS2-5637 : Parallel deployments

### PR DESCRIPTION
**Pull request by abhi-markan**
## Introduction
Due on-going parallel development on feature tickets and hot fixes, parallel deployment mechanism need has been identified.
Keeping this requirement in mind, existing CI/CD pipeline has been modified and improved.

* `dev` deployment will only push `main-amendment` branch changes, thus the pipeline finishes there.
* `staging` deployment will only push `main` branch changes, which once passed will then will be deployed to `production`.
* Amalgamation of deployment pipelines into one workflow with independent services as jobs.
* Removal of individual deployment pipelines.
* Workflow files have been renamed with consistent naming convention.
* Deployment script updated to `v0.0.3`.

### CODE REVIEW

File: .github/workflows/development_deploy.yml [hunk 1]
```diff
@@ -0,0 +1,640 @@
...
+        inlineScript: |
+          az webapp deployment slot create --name tfs-${{ env.environment }}-${{ env.repository }} --slot ${{ github.sha }} --configuration-source tfs-${{ env.environment }}-${{ env.repository }}
+          az webapp deployment slot swap --name tfs-${{ env.environment }}-${{ env.repository }} --slot ${{ github.sha }} --action swap
+          az webapp deployment slot delete --name tfs-${{ env.environment }}-${{ env.repository }} --slot ${{ github.sha }}
+
+    - name: 8. Restart
+      uses: azure/cli@v1.0.0
+      with:
+        inlineScript: |
+          az webapp restart --name tfs-${{ env.environment }}-${{ env.repository }}
```
Comment (IgnatG) [inline .github/workflows/development_deploy.yml:167]: this is nice :)

File: .github/workflows/staging_deploy.yml [hunk 2]
```diff
@@ -0,0 +1,640 @@
...
+          az webapp deployment slot swap --name tfs-${{ env.environment }}-${{ env.repository }} --slot ${{ github.sha }} --action swap
+          az webapp deployment slot delete --name tfs-${{ env.environment }}-${{ env.repository }} --slot ${{ github.sha }}
+
+    - name: 8. Restart
+      uses: azure/cli@v1.0.0
+      with:
+        inlineScript: |
+          az webapp restart --name tfs-${{ env.environment }}-${{ env.repository }}
+
+  #################### 6. DTFS API ####################
```
Comment (IgnatG) [inline .github/workflows/staging_deploy.yml:370]: same as the other file in dev

File: portal/README.md [hunk 3]
```diff
@@ -102,3 +102,4 @@ BSS could be seperated from Portal and the Portal UI's only purpose is to login
 3. Product pages served (/bss-ui or /gef-ui)
 
 # DTFS
+Trigger 12312312312312312
```
Comment (IgnatG) [inline portal/README.md:105]: we can remove this and replace it with a blank space if needed

**Review by IgnatG — COMMENTED** [review]
Beautiful! 
I've left some small comments

File: .github/workflows/deploy.sh [hunk 4]
```diff
@@ -25,47 +15,69 @@ printf "${RED}4. ACR Purge 🗑️${NC}\n\n"
...
+    branch=""
+    az acr run --cmd "acr purge --filter 'portal-ui:.*' --filter 'gef-ui:.*' --filter 'trade-finance-manager-ui:.*' --filter 'portal-api:.*' --filter 'trade-finance-manager-api:.*' --filter 'dtfs-central-api:.*' --filter 'reference-data-proxy:.*' --filter 'azure-function-number-generator:.*' --filter 'azure-function-acbs:.*' --ago 15d" --registry tfsdev /dev/null
+    az acr run --cmd "acr purge --filter 'portal-ui:.*' --filter 'gef-ui:.*' --filter 'trade-finance-manager-ui:.*' --filter 'portal-api:.*' --filter 'trade-finance-manager-api:.*' --filter 'dtfs-central-api:.*' --filter 'reference-data-proxy:.*' --filter 'azure-function-number-generator:.*' --filter 'azure-function-acbs:.*' --ago 15d" --registry tfsstaging /dev/null
+    ############### ACR PURGE ###############
+    fi
+
+    if [ -n "$destination" -a -n "$branch" ]
+    then
+    # Display latest push commit
+    git checkout ${branch}
```
Comment (IgnatG) [inline .github/workflows/deploy.sh:57]: do we need to use:
git checkout `"${branch}"`  ? 
we are getting some linting error on it.

File: .github/workflows/deploy.sh [hunk 5]
```diff
@@ -25,47 +15,69 @@ printf "${RED}4. ACR Purge 🗑️${NC}\n\n"
...
+    printf "${GREEN}"
+    git log -n 1 --pretty | sort | grep commit
+    printf "${NC}\n\n"
+
+    # Deploy
+    git checkout -b $destination
+    git push -f --set-upstream origin $destination
+
+    # Clean up
+    git checkout $branch
```
Comment (IgnatG) [inline .github/workflows/deploy.sh:69]: _same as above,_
do we need to use:
git checkout `"${branch}"`  ? 
we are getting some linting error on it.

File: .github/workflows/deploy.sh [hunk 6]
```diff
@@ -25,47 +15,69 @@ printf "${RED}4. ACR Purge 🗑️${NC}\n\n"
...
+    # Display latest push commit
+    git checkout ${branch}
+    git pull
+    printf "\n\n${NC}⬆️ ${branch} branch latest push : ${NC}"
+    printf "${GREEN}"
+    git log -n 1 --pretty | sort | grep commit
+    printf "${NC}\n\n"
+
+    # Deploy
+    git checkout -b $destination
```
Comment (IgnatG) [inline .github/workflows/deploy.sh:65]: same as the others - in quotes?

File: .github/workflows/development_deploy.yml [hunk 7]
```diff
@@ -0,0 +1,640 @@
+name: Development environment deployment
+
+on:
+  push:
+    branches: [ DTFS2-5637-review-the-deployment-pipelines-this-to-include-amendments ]
```
[the remark this answers is not in the document]
Comment (abhi-markan) [reply .github/workflows/development_deploy.yml]: Have just reverted it back, was awaiting pipeline to finish.

File: .github/workflows/development_deploy.yml [hunk 8]
```diff
@@ -0,0 +1,640 @@
...
+          az webapp deployment slot swap --name tfs-${{ env.environment }}-${{ env.repository }} --slot ${{ github.sha }} --action swap
+          az webapp deployment slot delete --name tfs-${{ env.environment }}-${{ env.repository }} --slot ${{ github.sha }}
+
+    - name: 8. Restart
+      uses: azure/cli@v1.0.0
+      with:
+        inlineScript: |
+          az webapp restart --name tfs-${{ env.environment }}-${{ env.repository }}
+
+  #################### 6. DTFS API ####################
```
[the remark this answers is not in the document]
Comment (abhi-markan) [reply .github/workflows/development_deploy.yml:370]: Added

File: .github/workflows/staging_deploy.yml (hunk 2 above)
Comment (abhi-markan) [reply .github/workflows/staging_deploy.yml:370]: Added

File: portal/README.md (hunk 3 above)
Comment (abhi-markan) [reply portal/README.md]: Removed

**Pull request merged**